File: test/embedded/Inputs/tiny-runtime-dummy-refcounting.c
#include <strings.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct ClassMetadata ClassMetadata;
typedef struct HeapObject HeapObject;

#if !__has_attribute(swiftcall)
#error "The runtime must be built with a compiler that supports swiftcall."
#endif

typedef struct ClassMetadata {
  ClassMetadata *superclassMetadata;
  void __attribute__((swiftcall)) (* destroy)(__attribute__((swift_context)) HeapObject *object);
} ClassMetadata;

typedef struct HeapObject {
  ClassMetadata *metadata;
  size_t refcount;
} HeapObject;

void *swift_slowAlloc(size_t bytes, size_t alignMask) {
  void *r = NULL;
  posix_memalign(&r, (alignMask == (size_t)-1) ? 16 : (alignMask + 1), bytes);
  bzero(r, bytes);
  return r;
}

void swift_slowDealloc(void *ptr, size_t bytes, size_t alignMask) {
  free(ptr);
}

void *swift_allocObject(ClassMetadata *metadata, size_t requiredSize, size_t requiredAlignmentMask) {
  HeapObject *object = swift_slowAlloc(requiredSize, requiredAlignmentMask);
  object->metadata = metadata;
  object->refcount = 1;
  return object;
}

void swift_deallocClassInstance(HeapObject *object, size_t allocatedSize, size_t allocatedAlignMask) {
  free(object);
}

HeapObject *swift_initStackObject(ClassMetadata *metadata, HeapObject *object) {
  object->metadata = metadata;
  object->refcount = -1;
  return object;
}

bool swift_isUniquelyReferenced_nonNull_native(HeapObject *object) {
  return object->refcount == 1;
}

void swift_release(HeapObject *object) {
  if (object == NULL) return;
  if (object->refcount == -1) return;

  object->refcount -= 1;
  if (object->refcount == 0) {
    object->metadata->destroy(object);
  }
}

HeapObject *swift_retain(HeapObject *object) {
  if (object->refcount == -1) return object;

  object->refcount += 1;
  return object;
}

void swift_beginAccess(void *pointer, void *buffer, uintptr_t flags, void *pc) { }
void swift_endAccess(void *buffer) { }

void swift_once(uintptr_t *predicate, void (*fn)(void *), void *context) {
  if (!*predicate) {
    *predicate = 1;
    fn(context);
  }
}

File: test/embedded/Inputs/tiny-runtime-dummy-refcounting.swift
let _swift_MinAllocationAlignment: UInt = 16

@_silgen_name("posix_memalign")
func posix_memalign(_: UnsafeMutablePointer<UnsafeMutableRawPointer?>, _: UInt, _: UInt) -> CInt

@_silgen_name("free")
func free(_: UnsafeMutableRawPointer?)

func alignedAlloc(size: UInt, alignment: UInt) -> UnsafeMutableRawPointer? {
  let alignment = max(alignment, UInt(MemoryLayout<UnsafeRawPointer>.size))
  var r: UnsafeMutableRawPointer? = nil
  _ = posix_memalign(&r, alignment, size)
  return r
}

/// Public APIs
// void *swift_slowAlloc(size_t size, size_t alignMask);
@_cdecl("swift_slowAlloc")
public func swift_slowAlloc(_ size: UInt, _ alignMask: UInt) -> UnsafeMutableRawPointer? {
  let alignment: UInt
  if alignMask == UInt.max {
    alignment = _swift_MinAllocationAlignment
  } else {
    alignment = alignMask + 1
  }
  return alignedAlloc(size: size, alignment: alignment)
}

// void swift_slowDealloc(void *ptr, size_t bytes, size_t alignMask);
@_cdecl("swift_slowDealloc")
public func swift_slowDealloc(_ ptr: UnsafeMutableRawPointer?, _ size: UInt, _ alignMask: UInt) {
  free(ptr)
}

@_silgen_name("swift_allocObject")
public func swift_allocObject(metadata: UnsafeMutableRawPointer, requiredSize: UInt, requiredAlignmentMask: UInt) -> UnsafeMutablePointer<HeapObject> {
  let p = swift_slowAlloc(requiredSize, requiredAlignmentMask)!
  let object = p.assumingMemoryBound(to: HeapObject.self)
  object.pointee.metadata = metadata
  object.pointee.refcount = 1
  return object
}

@_silgen_name("swift_deallocClassInstance")
public func swift_deallocClassInstance(object: UnsafeMutablePointer<HeapObject>, allocatedSize: UInt, allocatedAlignMask: UInt) {
  free(object)
}

@_silgen_name("swift_initStackObject")
public func swift_initStackObject(metadata: UnsafeMutableRawPointer, object: UnsafeMutablePointer<HeapObject>) -> UnsafeMutablePointer<HeapObject> {
  object.pointee.metadata = metadata
  object.pointee.refcount = -1
  return object
}

@_silgen_name("swift_isUniquelyReferenced_nonNull_native")
public func swift_isUniquelyReferenced_nonNull_native(object: UnsafeMutablePointer<HeapObject>) -> Bool {
  return object.pointee.refcount == 1
}

public struct HeapObject {
  var metadata: UnsafeMutableRawPointer
  var refcount: Int
}

@_silgen_name("swift_retain")
public func swift_retain(object: UnsafeMutablePointer<HeapObject>) -> UnsafeMutablePointer<HeapObject> {
  if object.pointee.refcount == -1 { return object }
  object.pointee.refcount += 1
  return object
}

@_silgen_name("swift_release")
public func swift_release(object: UnsafeMutablePointer<HeapObject>) {
  if object.pointee.refcount == -1 { return }
  object.pointee.refcount -= 1
  if object.pointee.refcount == 0 {
    free(object)
  }
}
